poseval/common/transformations.py

```diff
@@ -133,19 +183,55 @@ def _raw_board_to_sparse_representation(self, piece_to_layer_map, piece_to_value
                 values.append(piece_to_value_map[c])
                 column = column + 1
 
-        return ({'indices_z': indices_z, 'indices_x': indices_x, 'indices_y': indices_y, 'values': values})
+        return {'indices_z': indices_z, 'indices_x': indices_x, 'indices_y': indices_y, 'values': values}
 
     def _raw_board_to_flat_vector(self, piece_to_layer_map, piece_to_value_map, layers):
         raw_board = self.raw_board()
-        row = 0; column = 0
-        vector = [0 for _ in range(0, 8*8*layers)]
+        row = 0
+        column = 0
+        vector = [0 for _ in range(0, 8 * 8 * layers)]
         for c in raw_board:
             if c.isdigit():
                 column = column + int(c)
             elif c == '/':
                 row = row + 1
                 column = 0
             else:
-                vector[layers*(8*row + column) + piece_to_layer_map[c]] = piece_to_value_map[c]
+                vector[layers * (8 * row + column) + piece_to_layer_map[c]] = piece_to_value_map[c]
                 column = column + 1
         return vector
+
+    def castling_vector_8x8x2(self):
+        board = np.zeros([8, 8, 2])
+        board[(*DataSpecs.string_square_to_8x8('g8'), 1)] = self.castling_vector()[0]
+        board[(*DataSpecs.string_square_to_8x8('g1'), 0)] = self.castling_vector()[1]
+        board[(*DataSpecs.string_square_to_8x8('c8'), 1)] = self.castling_vector()[2]
+        board[(*DataSpecs.string_square_to_8x8('c1'), 0)] = self.castling_vector()[3]
+        return board
+
+    def en_passant_vector_8x8x1(self):
+        board = np.zeros([8, 8, 1])
+        if self.en_passant_target_square():
+            board[(*DataSpecs.string_square_to_8x8(self.en_passant_target_square()), 0)] = 1
+        return board
+
+    def full_board_representation(self, dim):
+        if dim == 6:
+            raise NotImplementedError  # TODO
+        return np.dstack(
+            (self.pieces_dense_representation(dim),
+             self.castling_vector_8x8x2(),
+             self.en_passant_vector_8x8x1()))
+
+    def pieces_dense_representation(self, dim):
+        board = np.zeros([8, 8, dim])
+        if dim == 6:
+            index_dict = self.raw_board_to_6x8x8_sparse_representation()
+        elif dim == 12:
+            index_dict = self.raw_board_to_12x8x8_sparse_representation()
+        else:
+            raise ValueError('Dimension not supported')
+        coordinates_list = zip(index_dict['indices_x'], index_dict['indices_y'], index_dict['indices_z'])
+        for coordinates, value in zip(coordinates_list, index_dict['values']):
```

Comment:
You sure that works? There's some black magic fuckery going on with `zip` in py3 iirc

Reply:
In fact, I'm zipping for clarity - first to make sure that coordinates are grouped as a tuple while iterating, and then to iterate jointly over values and coordinates.